File: src/test/diversion_sigtrap.py
from util import *
import re, sys

send_gdb('b breakpoint')
expect_gdb('Breakpoint 1')

send_gdb('c')
expect_gdb('Breakpoint 1')

send_gdb('p hardware_breakpoint()')
expect_gdb('Program received signal SIGTRAP')

File: src/test/diversion_sigtrap.c
/* -*- Mode: C; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */

#include "util.h"

static __attribute__((noinline)) void breakpoint(void) {
  int break_here = 1;
  (void)break_here;
}

static __attribute__((noinline)) int hardware_breakpoint(void) {
  asm("int3;");
  return 10;
}

int main(void) {
  breakpoint();
  hardware_breakpoint();
}
